File: BattleShipCSharp/SinglePlayerStateTracker.cs
using System;
using System.Collections.Generic;

namespace BattleShipCSharp
{
    public class SinglePlayerStateTracker
    {
        const int BoardSize = 10;
        public struct ShipSize
        {
            int size;
            private ShipSize(int _size) { size = _size; }
            public static implicit operator ShipSize(int size)
            {
                if (size < 1 || size > BoardSize)
                    throw new InvalidOperationException($"BattleShip size must be in the range [1..{BoardSize}]");
                return new ShipSize(size);
            }
            public static implicit operator int(ShipSize size) => size.size;
        }

        public enum CellStates { Water, Ship, Wreck }
        public enum ShipLayout { Horizontal, Vertical }
        public enum AttackResult { Miss, Hit, ReHit }

        CellStates[,] board;
        int shipCellsNbr, hitCellsNbr;

        public SinglePlayerStateTracker()
        {
            //no need to seed (default 0 = Water)
            board = new CellStates[BoardSize, BoardSize];
            shipCellsNbr = hitCellsNbr = 0;
        }

        public void AddBattleship(int leftX, int bottomY, ShipLayout layout, ShipSize size)
        {
            var rightX = leftX + (layout == ShipLayout.Horizontal ? size - 1 : 0);
            var topY = bottomY + (layout == ShipLayout.Vertical ? size - 1 : 0);

            if (leftX < 0 || bottomY < 0 || rightX >= BoardSize || topY >= BoardSize)
                throw new InvalidOperationException("The ship doesn't fit in the board");

            Func<int, int> X, Y;

            if (layout == ShipLayout.Horizontal)
            {
                X = (i) => leftX + i;
                Y = (_) => bottomY;
            }
            else
            {
                X = (_) => leftX;
                Y = (i) => bottomY + i;
            }

            var changedCells = new List<(int x, int y)>();

            try
            {
                for (int i = 0; i < size; i++)
                {
                    int x = X(i), y = Y(i);

                    if (board[x, y] != CellStates.Water)
                        throw new InvalidOperationException("The ship overlaps another ship's space");

                    board[x, y] = CellStates.Ship;
                    changedCells.Add((x, y));
                }
                shipCellsNbr += size;
            }
            catch
            {
                //rollback changes
                foreach (var (x, y) in changedCells)
                    board[x, y] = CellStates.Water;

                throw;
            }
        }

        public AttackResult Attack(int x, int y)
        {
            bool isOutOfBoard = x < 0 || y < 0 || x >= BoardSize || y >= BoardSize;
            CellStates cell = isOutOfBoard ? CellStates.Water : board[x, y];

            switch (cell)
            {
                case CellStates.Ship:
                    board[x, y] = CellStates.Wreck;
                    ++hitCellsNbr;
                    return AttackResult.Hit;

                case CellStates.Wreck:
                    return AttackResult.ReHit;

                default:
                    return AttackResult.Miss;
            }
        }

        public bool HasLost => hitCellsNbr >= shipCellsNbr;
    }
}

File: BattleShipFSharp/SinglePlayerStateTracker.fs
module BattleShipFSharp.SinglePlayerStateTracker

[<Literal>]
let BoardSize = 10

type CellStates = Water | Ship | Wreck
type ShipLayout = Horizontal | Vertical
type AttackResult = Miss | Hit | ReHit

type Game = private {
    Board: CellStates [,]
    ShipCellsNbr: int
    HitCellsNbr: int
}

type ShipSize = ShipSize of int
[<RequireQualifiedAccess>]
module private ShipSize = 
    let create raw = 
        if raw < 1 || raw > BoardSize then
            sprintf "BattleShip size must be in the range [1..%A]" BoardSize |> invalidOp
        else
            ShipSize raw

let start() = { Board = Array2D.init BoardSize BoardSize (fun _ _ -> Water); ShipCellsNbr = 0; HitCellsNbr = 0 }

let addBattleship leftX bottomY layout size game =
    let shipSize = ShipSize.create size
    
    let rightX = leftX + match layout with | Horizontal -> (size - 1) | _ -> 0
    let topY = bottomY + match layout with | Vertical -> (size - 1) | _ -> 0

    if ([leftX >= 0; bottomY >= 0; rightX < BoardSize; topY < BoardSize] |> List.exists not) then
        invalidOp "The ship doesn't fit in the board"

    let X, Y = 
        match layout with
        | Horizontal -> (fun i -> leftX + i), (fun _ -> bottomY)
        | Vertical -> (fun _ -> leftX), (fun i -> bottomY + i)

    let rec fitShip i changedCells =
        match shipSize with
        | ShipSize sz when i < sz ->
            let x, y = X(i), Y(i)
            match game.Board.[x, y] with
            | Water -> 
                game.Board.[x, y] <- Ship
                fitShip (i + 1) ((x, y) :: changedCells)
            | _ -> Error changedCells
        | _ -> Ok ()

    match fitShip 0 [] with
    | Ok _ -> { game with ShipCellsNbr = game.ShipCellsNbr + size }
    | Error changedCells -> 
        changedCells |> List.iter (fun (x, y) -> game.Board.[x, y] <- Water)
        invalidOp "The ship overlaps another ship's space"

let attack x y game =
    let isOutOfBoard = x < 0 || y < 0 || x >= BoardSize || y >= BoardSize
    let cell = if isOutOfBoard then Water else game.Board.[x, y]
    match cell with
    | Ship ->
        game.Board.[x, y] <- Wreck
        Hit, { game with HitCellsNbr = game.HitCellsNbr + 1 }
    | Wreck -> ReHit, game
    | Water -> Miss, game

let hasLost game = game.HitCellsNbr >= game.ShipCellsNbr
